# ChromBPNet: Deep learning models of base-resolution chromatin profiles
                This repo is under construction. Please check back in a week

- This repo contains code for the chrombpnet paper Anusri Pampari*, Anna Shcherbina*, Anshul Kundaje. (*authors contributed equally to the work)  
- Most have these scripts have been derived from an earlier repo (KerasAC) made by Anna Scherbina.
- General queries/thoughts have been addressed in the discussion section below.
- Please contact [Anusri Pampari] (\<first-name\>@stanford.edu)  for suggestions and comments. More instructions about reporting bugs detailed below.

## Quick Links

- [About](#chrombpnet)
- [Installation](#installation)
- [Tutorial](#tutorial-on-how-to-train-chrombpnet-models)
    - [Preprocessing](#preprocessing)
    - [Train and Evaluate Bias Model](#train-and-evaluate-bias-model)
    - [Train and Evaluate ChromBPNet Model](#train-and-evaluate-chrombpnet-model)
- [Variant Effect Prediction](#variant-effect-prediction)
- [Generate Genome Wide Browser Tracks](#generate-genome-wide-browser-tracks)
- [Invivo Footprinting](#invivo-footprinting)
- [Discussion](#discussion)
- [Report a bug](#chrombpnet-bugs)

##  ChromBPNet

Chromatin profiles (DNASE-seq and ATAC-seq) exhibit multi-resolution shapes and spans regulated by co-operative binding of transcription factors (TFs). This complexity is further difficult to mine because of confounding bias from enzymes (DNASE-I/Tn5) used in these assays. Existing methods do not account for this complexity at base-resolution and do not account for enzyme bias correctly, thus missing the high-resolution architecture of these profile. Here we introduce ChromBPNet to address both these aspects.

ChromBPNet (shown in the image as `chrombpnet model`) is a fully convolutional neural network that uses dilated convolutions with residual connections to enable large receptive fields with efficient parameterization. It also performs automatic assay bias correction in two steps, first by learning simple model on chromatin background that captures the enzyme effect (called `bias model` in the image). Then we use this model to regress out the effect of the enzyme from the ATAC-seq/Dnase-seq profiles. This two step process ensures that the sequence component of the ChromBPNet model (called `sequence model`) does not learn enzymatic bias. 

![Image](images/chrombpnet_arch.PNG)

If you are interested in learning more about the detailed architctures used, please refer to the following architecture files - 

- bias model: https://github.com/kundajelab/chrombpnet/blob/master/src/training/models/bpnet_model.py
- chrombpnet model: https://github.com/kundajelab/chrombpnet/blob/master/src/training/models/chrombpnet_model_with_bias_model.py.

## Installation

##  Tutorial on how to train chrombpnet models

Here we provide a step-by-step guide to training and evaluating chrombpnet models using the GM12878 ATAC-seq data (ENCSR095QNB) [here][url1]. This is bulk ATAC-seq data.

###  Preprocessing

#### Step 1: Download experimental data

We will first start by creating a directory named `data` and downloading the corresponding files (bams, overlap and idr peak sets) for ENCSR095QNB ENCODE dataset using the commans in the bash script below. 
```
bash step1_download_bams_and_peaks.sh

```

TODO - 
    - (define overlap and idr)
    - (When will we use overlap versus idr)
    - (Should bams be filtered or unfiltered)
    - (disuss how this will different for scATAC)


#### Step 2: Make Bigwigs from bam files (IMPORTANT STEP! PLEASE READ CAREFULLY)


We will now create unstranded bigwigs from the downloaded bam files (from step1) using the command below. This script uses the following two commands (1) `$PWD/src/helpers/preprocessing/bam_to_bigwig.sh`: Considers that the given bam files are *unshifted* and does a shift of +4/-4 and (2) `$PWD/src/helpers/preprocessing/analysis/build_pwm_from_bigwig.py` generates an image of the bias motif. 
```
bash step2_make_bigwigs_from_bams.sh data/merged.bam data/ ATAC_PE data/hg38.fa data/hg38.chrom.sizes
```

After running this command open the `bias_pwm.png` image generated by the script in this folder. You will see the following tn5 motif PWM.

![Image](images/bias_pwm.png)

*IMPORTANT NOTE 1:* If you are running these commands on custom experimental bam - *read the documentation* in the directory `$PWD/src/helpers/preprocessing/` to make sure you are using the script correctly. After that use this script `$PWD/src/helpers/preprocessing/analysis/` and making sure you see the tn5 bias pwm. If you do not see this it is likely the bam's that you provided have a shift of some kind. Please provide only unshifted bams to the script. *Do not proceed further if you do not see a tn5 motif after this step.* 

*IMPORTANT NOTE 2:* If you are running the pipeline on custom generarated bigwigs (without using `$PWD/src/helpers/preprocessing/`) make sure the bigwigs are unstranded and make sure the shifts are done correctly. To check this `$PWD/src/helpers/preprocessing/analysis/` run the scripts in this directory and make sure you see the tn5 pwm. *Do not proceed further if you do not see a tn5 motif after this step.* If you see another kind of motif as shown here - you have the incorrect shift. go back and fix this.

TODO - 
    - (define unstranded)
    - write analysis script to tell yes or no

#### Step 3: Step 3: Generate background regions gc-matched with the peaks

Here we will generate non-peak background regions that gc-match with the peak regions. We will use thee regions to train and evaluate a bias model. We will also use these regions as bacjground regions tos get marginal footprints.

First we will start by dividing the entire genome into overlapping bins of `inputlen` regions. ChromBpnet models are trained on `inputlen` of 2114 on human genome datasets. This value was chosen such that all chromatin features are captures and performance saturates. This step takes several hours (Sorry never got to speeding this step sinece its a one-time step, if you have ways to speed this step please contribute.) For conveinve the human genome data we use is provided here - . But if you want to make this file from scratch follow these steps


Now we will find the regions from this genome-wide bucket that do not fall in overlap-peaks or blacklist bed but have same gc-disribution as the overlap-peak set

```
bash step3_get_background_regions.sh data/merged.bam data/ ATAC_PE data/hg38.fa data/hg38.chrom.sizes
```


###  Train and Evaluate Bias Model

We are now ready to train a bias model! 


#### Step 4: Train Bias Model


#### Step 5: Interpret bias model


#### Step 6: Scale bias model


###  Train and Evaluate ChromBPNet Model


#### Step 7: Train ChromBPNet Model (This step will also generate the sequence model)


#### Step 8: Interpret chrombpnet model and sequence model


#### Step 9: Marginal footprinting using bias, chrombpnet ans sequence model


##  Variant Effect Prediction

## Generate Genome Wide Browser Tracks


##  Discusssion


##  ChromBPNet bugs


[url1]: https://www.encodeproject.org/experiments/ENCSR095QNB/
